orttraining/orttraining/models/bert/main.cc

```diff
@@ -743,10 +743,10 @@ static Status RunTraining(const BertParameters& params, const Environment& env)
     ORT_RETURN_IF_ERROR(runner->ResetLossScaler());
   }
 
-  auto test_data_loader = onnxruntime::make_unique<DataLoader>(params_for_phase.input_name_map,
-                                                               params_for_phase.test_data_dir,
-                                                               max_num_files_preload);
-  ORT_RETURN_IF_ERROR(runner->EndTraining(test_data_loader.get()));
+  // only test and save trained model on device #0
+  if (params_for_phase.mpi_context.world_rank == 0) {
+    ORT_RETURN_IF_ERROR(runner->EndTraining(nullptr));
```

Comment:
so the hang issue only happens on when tensorboard enabled. Shall we add comment saying something like this?
"TensorBoard operator SummaryMerge in the graph requires some computation after NcclAllReduce. So running this only on rank 0, make all reduce hangs there. "

Reply:
Yes, I think this is more detailed. I will update.

Reply:
oh actually the fix is: to pass nullptr as test data, so this will disable evaluation. Running only on rank 0 is the reason Reduce hangs, since worker 0 is waiting for other workers but they did not response.